Task_1_3_1

MIZINCHIK opened the pull request.

MIZINCHIK committed:
Added basics

MIZINCHIK committed:
Fix building

MIZINCHIK committed:
Fix building

MIZINCHIK committed:
Fix gradle

MIZINCHIK committed:
Create SubstringFinderTest.java

MIZINCHIK committed:
write algo

added a fully working algo with some tests

MIZINCHIK committed:
added specific tests

MIZINCHIK committed:
add large file test

MIZINCHIK committed:
fix large test

MIZINCHIK committed:
add javadoc

MIZINCHIK committed:
delete cyrillic test

escape characters are not allowed thus cannot perform a reference test

MIZINCHIK committed:
fix reviewdog

MIZINCHIK committed:
fix reviewdog

alradish commented on Task_1_3_1/src/main/java/io/github/mizinchik/SubstringFinder.java:
```diff
@@ -0,0 +1,45 @@
+     *
+     * @param reader source of text
+     */
+    void eatReader(Reader reader);
```
Результат нужно вернуть тому кто вызвал метод, и он уже сам будет решать, что с этим делать

alradish commented on Task_1_3_1/src/main/java/io/github/mizinchik/SubstringFinderImpl.java:
```diff
@@ -0,0 +1,204 @@
+     */
+    private void buildZarray() throws UnsupportedOperationException {
+        if (reader == null || substring == null) {
+            throw new UnsupportedOperationException("Can't build a Z array w/o"
```
Это скорее IllegalStateException, в смысле, что состояние вашего объекта не подходит для выполнения. Чаще всего ансапортед используют когда вообще не реализовали метод

alradish commented on Task_1_3_1/src/main/java/io/github/mizinchik/SubstringFinderImpl.java:
```diff
@@ -0,0 +1,204 @@
+        if (reader == null || substring == null) {
+            throw new UnsupportedOperationException("Can't build a Z array w/o"
+                    + " a reader and/or a substring");
+        } else {
```
else лишний. Так как если вы упали в ветку ифа, то вы уже вышли из метода

alradish commented on Task_1_3_1/src/main/java/io/github/mizinchik/SubstringFinderImpl.java:
```diff
@@ -0,0 +1,204 @@
+                int intLeftIndex = 0;
+                int intRightIndex = 0;
+                stringArray.add(0, 0);
+                for (int i = 1; i < substring.length(); i++) {
+                    if (i > intRightIndex) {
+                        intLeftIndex = i;
+                        intRightIndex = i;
+                        while (intRightIndex < substring.length()
+                                && substring.charAt(intRightIndex - intLeftIndex)
+                                == substring.charAt(intRightIndex)) {
+                            intRightIndex++;
+                        }
+                        stringArray.add(i, intRightIndex - intLeftIndex);
+                        intRightIndex--;
+                    } else {
+                        int insideIndex = i - intLeftIndex;
+                        if (stringArray.get(insideIndex) < (intRightIndex - i + 1)) {
+                            stringArray.add(i, stringArray.get(insideIndex));
+                        } else {
+                            intLeftIndex = i;
+                            while (intRightIndex < substring.length()
+                                    && substring.charAt(intRightIndex - intLeftIndex)
+                                    == substring.charAt(intRightIndex)) {
+                                intRightIndex++;
+                            }
+                            stringArray.add(i, intRightIndex - intLeftIndex);
+                            intRightIndex--;
+                        }
+                    }
+                }
```
Дублирование кода. Напишите удобные метод вычисления z функции и используйте его здесь и ниже

MIZINCHIK committed:
fix pr issues

MIZINCHIK committed:
fix errors and reviewdog

MIZINCHIK committed:
return big file

MIZINCHIK committed:
make different z-functions useful

make an existence of 2 ways of finding a z-array actually useful inside a program

MIZINCHIK committed:
delete excess code

alradish commented on Task_1_3_1/src/main/java/io/github/mizinchik/SubstringFinderImpl.java:
```diff
@@ -0,0 +1,245 @@
+    private String substring;
+    private Reader reader;
+    private final Map<Long, Integer> zarray;
+    List<Integer> stringList;
```
private

alradish commented on Task_1_3_1/src/main/java/io/github/mizinchik/SubstringFinderImpl.java:
```diff
@@ -0,0 +1,245 @@
+    public SubstringFinderImpl(String fileName, String substring)
+            throws IllegalStateException, FileNotFoundException {
+        this.substring = substring;
+        this.reader = new InputStreamReader(new FileInputStream(fileName), StandardCharsets.UTF_8);
```
Надо бы закрыть. Используйте try with resources

alradish reviewed (changes requested):
От дублирования кода нужно избавиться. Это критично

MIZINCHIK committed:
fix code duplication

+ drop large test

MIZINCHIK committed:
fix import

MIZINCHIK committed:
fix string length

MIZINCHIK committed:
fix variable name

MIZINCHIK committed:
fix string length

MIZINCHIK committed:
fix lexicographical order

MIZINCHIK committed:
bring large test back
